src/libappimage/core/impl/TraversalType2.cpp

```diff
@@ -0,0 +1,296 @@
+
+TraversalType2::TraversalType2(std::string path) : path(path) {
+    clog << "Opening " << path << " as Type 2 AppImage" << endl;
+    // read the offset at which a squashfs image is expected to start
+    ssize_t fs_offset = core::AppImage(path).getPayloadOffset();
+
+    if (fs_offset < 0)
+        throw AppImageReadError("get_elf_size error");
+
+    sqfs_err err = sqfs_open_image(&fs, path.c_str(), (size_t) fs_offset);
+
+    if (err != SQFS_OK)
+        throw AppImageReadError("sqfs_open_image error: " + path);
+
+    // prepare for traverse
+    rootInodeId = sqfs_inode_root(&fs);
+    err = sqfs_traverse_open(&trv, &fs, rootInodeId);
+    if (err != SQFS_OK)
+        throw AppImageReadError("sqfs_traverse_open error");
+
+}
+
+TraversalType2::~TraversalType2() {
+    sqfs_traverse_close(&trv);
+
+    clog << "Closing " << path << " as Type 2 AppImage" << endl;
+    sqfs_destroy(&fs);
+}
+
+void TraversalType2::next() {
+    sqfs_err err;
+    if (!sqfs_traverse_next(&trv, &err))
+        completed = true;
+
+    if (err != SQFS_OK)
+        throw AppImageReadError("sqfs_traverse_next error");
+}
+
+bool TraversalType2::isCompleted() {
+    return completed;
+}
+
+std::string TraversalType2::getEntryName() {
+    if (trv.path != nullptr)
+        return trv.path;
+    else
+        return string();
+}
+
+appimage::core::entry::Type TraversalType2::getEntryType() {
+    if (trv.dir_end)
+        return entry::DIR;
+
+    switch (trv.entry.type) {
+        case SQUASHFS_REG_TYPE:
+        case SQUASHFS_LREG_TYPE:
+            return entry::REGULAR;
+
+        case SQUASHFS_SYMLINK_TYPE:
+        case SQUASHFS_LSYMLINK_TYPE:
+            return entry::LINK;
+
+        case SQUASHFS_DIR_TYPE:
+        case SQUASHFS_LDIR_TYPE:
+            return entry::DIR;
+
+        default:
+            return entry::UNKNOWN;
+    }
+}
+
+void TraversalType2::extract(const std::string& target) {
+    sqfs_inode inode;
+    if (sqfs_inode_get(&fs, &inode, trv.entry.inode))
+        throw AppImageReadError("sqfs_inode_get error");
+
+    // create target parent dir
+    auto parentPath = boost::filesystem::path(target).parent_path();
+    boost::filesystem::create_directories(parentPath);
+
+    // handle each inode type properly
+    switch (inode.base.inode_type) {
+        case SQUASHFS_DIR_TYPE:
+        case SQUASHFS_LDIR_TYPE:
+            extractDir(target);
+            break;
+        case SQUASHFS_REG_TYPE:
+        case SQUASHFS_LREG_TYPE:
+            extractFile(inode, target);
+            break;
+        case SQUASHFS_SYMLINK_TYPE:
+        case SQUASHFS_LSYMLINK_TYPE:
+            extractSymlink(inode, target);
+            break;
+        default:
+            throw AppImageError("AppImage Type 2 inode.base.inode_type " +
+                                std::to_string(inode.base.inode_type) + " not supported yet.");
+    }
+}
+
+sqfs_err TraversalType2::sqfs_stat(sqfs* fs, sqfs_inode* inode, struct stat* st) {
```

Comment:
`TraversalType2` is a private class, no need to use PImpl here.

Reply:
Please try to group such classes together (not by name but in actual modules, e.g., in directories which define `OBJECT` libraries which can be included in a main library) in this case. Otherwise, it'll just look super inconsistent when a pattern like PImpl is used inconsequently.

Reply:
Originally I wanted to put the contents of `src/libappimage/core/impl` into a OBJECT library but the fact that combining two OBJECT libraries is not possible on cmake 3.2 ( AFIK, not even in the latest 3.12)